core/src/main/java/org/elasticsearch/cluster/service/TaskBatching.java

```diff
@@ -0,0 +1,233 @@
+            for (BatchingTask task : updateTasks) {
+                if (task.processed.getAndSet(true) == false) {
+                    logger.debug("task [{}] timed out after [{}]", task.source, timeout);
+                    toRemove.add(task);
+                }
+            }
+            if (toRemove.isEmpty() == false) {
+                Object batchingExecutor = toRemove.get(0).batchingKey;
+                synchronized (tasksPerExecutor) {
+                    LinkedHashSet<BatchingTask> existingTasks = tasksPerExecutor.get(batchingExecutor);
+                    if (existingTasks != null) {
+                        existingTasks.removeAll(toRemove);
+                        if (existingTasks.isEmpty()) {
+                            tasksPerExecutor.remove(batchingExecutor);
+                        }
+                    }
+                }
+                for (BatchingTask task : toRemove) {
+                    onTimeout(task, timeout);
+                }
+            }
+        });
+    }
+
+    /**
+     * Action to be implemented by the specific batching implementation
+     */
+    protected abstract void onTimeout(BatchingTask task, TimeValue timeout);
+
+    void runIfNotProcessed(BatchingTask updateTask) {
+        // if this task is already processed, the executor shouldn't execute other tasks (that arrived later),
+        // to give other executors a chance to execute their tasks.
+        if (updateTask.processed.get() == false) {
+            final List<BatchingTask> toExecute = new ArrayList<>();
+            final Map<String, List<BatchingTask>> processTasksBySource = new HashMap<>();
+            synchronized (tasksPerExecutor) {
+                LinkedHashSet<BatchingTask> pending = tasksPerExecutor.remove(updateTask.batchingKey);
+                if (pending != null) {
+                    for (BatchingTask task : pending) {
+                        if (task.processed.getAndSet(true) == false) {
+                            logger.trace("will process {}", task);
+                            toExecute.add(task);
+                            processTasksBySource.computeIfAbsent(task.source, s -> new ArrayList<>()).add(task);
+                        } else {
+                            logger.trace("skipping {}, already processed", task);
+                        }
+                    }
+                }
+            }
+
+            if (toExecute.isEmpty() == false) {
+                final String tasksSummary = processTasksBySource.entrySet().stream().map(entry -> {
+                    String tasks = updateTask.describeTasks(entry.getValue());
+                    return tasks.isEmpty() ? entry.getKey() : entry.getKey() + "[" + tasks + "]";
+                }).reduce((s1, s2) -> s1 + ", " + s2).orElse("");
+
+                run(updateTask.batchingKey, toExecute, tasksSummary);
+            }
+        }
+    }
+
+    /**
+     * Action to be implemented by the specific batching implementation
```

Comment:
can we add a comment that all the tasks are guaranteed to have to supplied batching key?